### ISSUE TITLE
List<String> generated from ResultSet show inconsistent marks as a List<String> written by hand

### ISSUE THREAD

**Issue by akon89** [wumpz/java-diff-utils#18]
### Expected Behavior
ResultSet from JDBC converted into Strings and then added to a List<String> should show the same differences as a List<String> = Arrays.asList() created by hand.

When using this (entered by hand)
```
List<String> listOne = Arrays.asList("MASS_TABLE,ID,NUMBER,22,N,", "MASS_TABLE,ISSUEID,NUMBER,22,Y,", "MASS_TABLE,MODIFIED,NUMBER,22,Y,", "MASS_TABLE,SYSTEMNAME,VARCHAR2,1020,Y,");
List<String> listTwo = Arrays.asList("ATOM_TABLE,ID,NUMBER,22,N,", "ATOM_TABLE,ISSUEID,NUMBER,22,Y,", "ATOM_TABLE,MODIFIED,NUMBER,22,Y,", "ATOM_TABLE,ACRONYM,VARCHAR2,255,Y," );
```

You will get this when ran through the DiffGenerator:

```
|~MASS_TABLE~,ID,NUMBER,22,N, | **ATOM_TABLE**,ID,NUMBER,22,N,|
|~MASS_TABLE~,ISSUEID,NUMBER,22,Y, | **ATOM_TABLE**,ISSUEID,NUMBER,22,Y,|
|~MASS_TABLE~,MODIFIED,NUMBER,22,Y, | **ATOM_TABLE**,MODIFIED,NUMBER,22,Y,|
|~MASS_TABLE~,~SYSTEMNAME~,VARCHAR2,~1020~,Y, | **ATOM_TABLE**,**ACRONYM**,VARCHAR2,**255**,Y,|
```
### Actual Behavior
DiffGenerator shows apparently incorrect results.

Example:

("NUMBER" on the first row, and "VARCHAR" on the last row show marks ("**") in front of the word to indicate the difference, but don't show the corresponding marks behind it).  Also, as the results go further, some differences will be marked on the first list and no corresponding marks will be shown at all on the second list.   

```
| ~MASS_TABLE~,ID,NUMBER,22,N, | **ATOM_TABLE**,**ID**,**NUMBER,22,N, |
| ~MASS_TABLE~,ISSUEID,NUMBER,22,Y, | ATOM_TABLE,ISSUEID,NUMBER,22,Y, |
| ~MASS_TABLE~,MODIFIED,NUMBER,22,Y, | ATOM_TABLE,MODIFIED,NUMBER,22,Y, |
| ~MASS_TABLE~,SYSTEMNAME,VARCHAR2,~1020~,Y, | ATOM_TABLE,ACRONYM,**VARCHAR2,**255**,Y, |
```

### Steps to Reproduce the Problem

  1.Connect to two different Oracle Databases and run a SQL query, saving information as text from 
     ResultSet into different List<String>
  2.Compare the two Lists in the DiffGenerator
  3.Compare these results to the results when the information is put in by hand and compared using DiffGenerator.

Shows as expected:
```
  List<String> listOne = Arrays.asList("MASS_TABLE,ID,NUMBER,22,N,", "MASS_TABLE,ISSUEID,NUMBER,22,Y,", "MASS_TABLE,MODIFIED,NUMBER,22,Y,", "MASS_TABLE,SYSTEMNAME,VARCHAR2,1020,Y,");
        List<String> listTwo = Arrays.asList("ATOM_TABLE,ID,NUMBER,22,N,", "ATOM_TABLE,ISSUEID,NUMBER,22,Y,", "ATOM_TABLE,MODIFIED,NUMBER,22,Y,", "ATOM_TABLE,ACRONYM,VARCHAR2,255,Y," );

        DiffRowGenerator generator = DiffRowGenerator.create()
                .showInlineDiffs(true)
                .inlineDiffByWord(true)
                .oldTag(f -> "~")
                .newTag(f -> "**") 
                .build();

        List<DiffRow> diffRowList;

        try {
            diffRowList = generator.generateDiffRows(listOne, listTwo);

            for (DiffRow diffRow : diffRowList) {
                System.out.println("|" + diffRow.getOldLine() + "|" + diffRow.getNewLine() + "|");
            }
        } catch (DiffException e) {
            e.printStackTrace();
        }
```

Does not show as expected, using database:
```
...
  List<String> databaseSchemaList = new ArrayList<>();
        try {
            conn = DBHelper.getConnection(databaseName);
            stmt = Objects.requireNonNull(conn).createStatement();
            rs = stmt.executeQuery(query);

            while (rs.next()) {
                int columnCount = rs.getMetaData().getColumnCount();
                StringBuilder rsStringBuilder = new StringBuilder();
                for (int i = 1; i <= columnCount; i++) {
                    Object rsObject = rs.getObject(i);
                    String rsString = (rsObject == null) ? "NULL" : rsObject.toString();
                    rsStringBuilder.append(rsString).append(",");
                }         
                databaseSchemaList.add(rsStringBuilder.toString());
            }
       
        } catch (Exception e) {
            e.printStackTrace();
        } finally {     
            ...
        }
        return databaseSchemaList;   
    }

    private void getComparisonResults(List<String> listOne, List<String> listTwo) {

        DiffRowGenerator generator = DiffRowGenerator.create()
                .showInlineDiffs(true)
                .inlineDiffByWord(true)
                .oldTag(f -> "~") 
                .newTag(f -> "**")
                .build();

        List<DiffRow> diffRowList;

        try {
            diffRowList = generator.generateDiffRows(listOne, listTwo);

            for (DiffRow diffRow : diffRowList) {
                System.out.println("| " + diffRow.getOldLine() + " | " + diffRow.getNewLine() + " |");
            }
      ...
}

 public static void main(String[] args) {

        SchemaComparison schemaComparison = new SchemaComparison();

        //get list from each database
        List<String> aSchemaList = schemaComparison.getDatabaseSchemaList(DatabaseName.A);
        List<String> bSchemaList = schemaComparison.getDatabaseSchemaList(DatabaseName.B);

       //run comparison
        schemaComparison.getComparisonResults(aSchemaList,bSchemaList);

    }

```

### Specifications

  - Version: 2.3 SNAPSHOT
  - Platform: Windows 7
  - Subsystem: Oracle Database 11g, IntelliJ 2018.1

**Comment by wumpz** [comment]
I made a test using Oracle and could not reproduce your problem. I am not sure which version of diffutils I deployed. Sure it is 2.3-SNAPSHOT, but I am not sure how old it is. Could you rebuild from sources and test this?

**Comment by wumpz** [comment]
Could you give me your database generated values as bytes? There we could check for special characters.

**Comment by akon89** [comment]
Here are some values from the first database (from System.out.println):
> [B@3e2e18f2
> [B@470f1802
> [B@63021689
> [B@703580bf
> [B@3e92efc3
> [B@1622f1b
> [B@72a7c7e0
> [B@2e4b8173
> [B@70e8f8e
> [B@17046283
> [B@5bd03f44
> [B@29626d54
> [B@5a63f509
> [B@6e4784bc
> [B@34b7ac2f
> [B@e056f20
> [B@4b0b0854
> [B@19bb07ed
> [B@10e41621
> [B@353d0772
> [B@2667f029
> [B@67a20f67
> null
> [B@57c758ac
> [B@a9cd3b1
> [B@13e39c73
> [B@64cd705f
> [B@9225652
> null
> [B@654f0d9c
> [B@6a400542
> [B@6580cfdd
> [B@7e0b85f9
> [B@63355449
> [B@9353778

Second:
> [B@80b122b
> [B@999b951
> [B@65cf4d6d
> [B@1b410308
> null
> [B@4b0bd001
> [B@f9cab00
> [B@10272bbb
> [B@36d33f4
> [B@16736040
> null
> [B@7c5d1d25
> [B@550e9be6
> [B@3f96f020
> [B@32eae6f2
> [B@26ca61bf
> null
> [B@73a0f2b
> [B@fa5f81c
> [B@6a1d526c
> [B@1f2f0109
> [B@483b0690
> null
> [B@687e6293
> [B@6870c3c2
> [B@fb0a08c
> [B@1faf386c
> [B@4debbf0
> [B@6a5e167a
> [B@60e06f7d

**Comment by wumpz** [comment]
You are sending not the bytes, but the hash values of your byte arrays instances.

`System.out.println(bytes); `is the problem. Better use something like:

```Java
System.out.println(
    new BigInteger(bytes).toString(16)
); 
```

This is a neat little trick to get a hex ouput without using any external library or special methods.

**Comment by akon89** [comment]
Thank you for explaining that. 
```
   if (bytes != null) {
                        System.out.println(
                                new BigInteger(bytes).toString(16)
                        );
                    }
                    else{
                        System.out.println("byte value was null");
                    }
                }
```
First:
> 414354494f4e535f433137303035
> 4944
> 4e554d424552
> -3ee9
> -3eec
> 4e
> 414354494f4e535f433137303035
> 49535355454944
> 4e554d424552
> -3ee9
> -3eec
> 59
> 414354494f4e535f433137303035
> 4d4f444946494544
> 4e554d424552
> -3ee9
> -3ef5
> 59
> 414354494f4e535f433137303035
> 4143524f4e594d
> 5641524348415232
> -3df4eb
> byte value was null
> 59
> 414354494f4e535f433137303035
> 4143524f4e594d5f444546494e4954494f4e
> 434c4f42
> -3dd7
> byte value was null
> 59
> 414354494f4e535f433137303038
> 4944
> 4e554d424552
> -3ee9
> -3eec
> 4e
> 414354494f4e535f433137303038
> 49535355454944
> 4e554d424552
> -3ee9
> -3eec
> 59
> 414354494f4e535f433137303038
> 4d4f444946494544
> 4e554d424552
> -3ee9
> -3ef5
> 59
> 414354494f4e535f433137303038
> 574152
> 4e554d424552
> -3ee9
> -3efe
> 59
> 414354494f4e535f433137303038
> 4950524f4a454354
> 5641524348415232
> -3df4eb
> byte value was null
> 59
> 414354494f4e535f433137303038
> 4950524f4a4543545f4e414d45
> 5641524348415232
> -3df4eb
> byte value was null

Second:

> 414354494f4e535f433136393133
> 5553554d4d415259
> 5641524348415232
> -3dfcc8
> byte value was null
> 59
> 414354494f4e535f433136393133
> 49445545
> 44415445
> -3ef8
> byte value was null
> 59
> 414354494f4e535f433136393133
> 554445534352495054494f4e
> 5641524348415232
> -3dfcc8
> byte value was null
> 59
> 414354494f4e535f433136393133
> 5544415445
> 44415445
> -3ef8
> byte value was null
> 59
> 414354494f4e535f433136393133
> 4944
> 4e554d424552
> -3ee9
> -3eec
> 4e
> 414354494f4e535f433136393133
> 49535355454944
> 4e554d424552
> -3ee9
> -3eec
> 59
> 414354494f4e535f433136393133
> 4d4f444946494544
> 4e554d424552
> -3ee9
> -3ef5
> 59
>

**Comment by akon89** [comment]
The information comes from the SQL query that gets information about all the tables that the user has access to.  The output from the query gives back around a couple thousand strings for each database, because of all the tables it contains, which would be too much to post here, so I just shortened some of it. The errors are similar to what I posted above, but after a few hundred lines there are some that look like the comparison is working correctly, but then later it looks incorrect again.  
I think the table names I gave you earlier no longer exist but similar ones were recreated and they show the same problems, and this is the same query.
This is the current query I have been using:

```
SELECT
  TABLE_NAME,
  COLUMN_NAME,
  DATA_TYPE,
  DATA_LENGTH,
  DATA_PRECISION,
  NULLABLE
FROM all_tab_columns
WHERE OWNER = 'USER'
ORDER BY TABLE_NAME ASC; 
```

**Comment by wumpz** [comment]
I did wanted your sql, but the concrete values. You see if I run your sql or use your primary test data then this negatives should be the precision of your columns, e.g. 22. But 22 is encoded to hex *3232* and not negative.
So the problem seems to be in this rows.

**Comment by akon89** [comment]
Ok, I can provide you some values directly.  When I run the query in DataGrip then  I get these back (the first 36 rows)
First
```
TABLE_NAME,COLUMN_NAME,DATA_TYPE,DATA_LENGTH,DATA_PRECISION,NULLABLE
ACTIONS_C17005,ID,NUMBER,22,19,N
ACTIONS_C17005,ISSUEID,NUMBER,22,19,Y
ACTIONS_C17005,MODIFIED,NUMBER,22,10,Y
ACTIONS_C17005,ACRONYM,VARCHAR2,1020,,Y
ACTIONS_C17005,ACRONYM_DEFINITION,CLOB,4000,,Y
ACTIONS_C17008,ID,NUMBER,22,19,N
ACTIONS_C17008,ISSUEID,NUMBER,22,19,Y
ACTIONS_C17008,MODIFIED,NUMBER,22,10,Y
ACTIONS_C17008,WAR,NUMBER,22,1,Y
ACTIONS_C17008,IPROJECT,VARCHAR2,1020,,Y
ACTIONS_C17008,IPROJECT_NAME,VARCHAR2,1020,,Y
ACTIONS_C17008,TASK,VARCHAR2,1020,,Y
ACTIONS_C17008,HOURS_PROJECTED,VARCHAR2,1020,,Y
ACTIONS_C17008,HOURS_SPENT,VARCHAR2,1020,,Y
ACTIONS_C17008,START_DATE,DATE,7,,Y
ACTIONS_C17008,COMPLETE_DATE,DATE,7,,Y
ACTIONS_C17008,ISTATUS,VARCHAR2,1020,,Y
ACTIONS_C17008,ISTATUS_NAME,VARCHAR2,1020,,Y
ACTIONS_C18527,VERIFICAITONTYPE,VARCHAR2,1020,,Y
ACTIONS_C18527,ID,NUMBER,22,19,N
ACTIONS_C18527,ISSUEID,NUMBER,22,19,Y
ACTIONS_C18527,MODIFIED,NUMBER,22,10,Y
ACTIONS_C18527,VERIFICATIONTYPE,VARCHAR2,1020,,Y
ACTIONS_C18527,VERIFICATIONNUMBER,VARCHAR2,1020,,Y
ACTIONS_C18539,USUMMARY,VARCHAR2,1020,,Y
ACTIONS_C18539,IASSIGNEE,VARCHAR2,1020,,Y
ACTIONS_C18539,IASSIGNEE_NAME,VARCHAR2,1020,,Y
ACTIONS_C18539,ISTATUS,VARCHAR2,1020,,Y
ACTIONS_C18539,ISTATUS_NAME,VARCHAR2,1020,,Y
ACTIONS_C18539,IDUE,DATE,7,,Y
ACTIONS_C18539,URL,VARCHAR2,1020,,Y
ACTIONS_C18539,ID,NUMBER,22,19,N
ACTIONS_C18539,ISSUEID,NUMBER,22,19,Y
ACTIONS_C18539,MODIFIED,NUMBER,22,10,Y
ACTIONS_C18539,DOCUMENTNAME,VARCHAR2,1020,,Y
ACTIONS_C18539,DOCUMENTLOCATION,VARCHAR2,1020,,Y
```
Second
```
TABLE_NAME,COLUMN_NAME,DATA_TYPE,DATA_LENGTH,DATA_PRECISION,NULLABLE
ACTIONS_C16913,USUMMARY,VARCHAR2,255,,Y
ACTIONS_C16913,IDUE,DATE,7,,Y
ACTIONS_C16913,UDESCRIPTION,VARCHAR2,255,,Y
ACTIONS_C16913,UDATE,DATE,7,,Y
ACTIONS_C16913,ID,NUMBER,22,19,N
ACTIONS_C16913,ISSUEID,NUMBER,22,19,Y
ACTIONS_C16913,MODIFIED,NUMBER,22,10,Y
ACTIONS_C16913,WAR,NUMBER,22,1,Y
ACTIONS_C16913,IPROJECT,VARCHAR2,255,,Y
ACTIONS_C16913,IPROJECT_NAME,VARCHAR2,255,,Y
ACTIONS_C16913,TASK,VARCHAR2,255,,Y
ACTIONS_C16913,HOURS_SPENT,VARCHAR2,255,,Y
ACTIONS_C16913,START_DATE,DATE,7,,Y
ACTIONS_C16913,COMPLETE_DATE,DATE,7,,Y
ACTIONS_C16913,ISTATUS,VARCHAR2,255,,Y
ACTIONS_C16913,ISTATUS_NAME,VARCHAR2,255,,Y
ACTIONS_C16913,RELEASE_DATE,DATE,7,,Y
ACTIONS_C16913,DEV_DATE,DATE,7,,Y
ACTIONS_C16913,PROD_DATE,DATE,7,,Y
ACTIONS_C16913,UAT_DATE,DATE,7,,Y
ACTIONS_C16913,HOURS_PROJECTED,VARCHAR2,255,,Y
ACTIONS_C16913,IASSIGNEE,VARCHAR2,255,,Y
ACTIONS_C16913,IASSIGNEE_NAME,VARCHAR2,255,,Y
ACTIONS_C16913,LOG_ITEM,VARCHAR2,255,,Y
ACTIONS_C16915,ID,NUMBER,22,19,N
ACTIONS_C16915,ISSUEID,NUMBER,22,19,Y
ACTIONS_C16915,MODIFIED,NUMBER,22,10,Y
ACTIONS_C16915,WAR,NUMBER,22,1,Y
ACTIONS_C16915,IPROJECT,VARCHAR2,255,,Y
ACTIONS_C16915,IPROJECT_NAME,VARCHAR2,255,,Y
ACTIONS_C16915,TASK,VARCHAR2,255,,Y
ACTIONS_C16915,HOURS_SPENT,VARCHAR2,255,,Y
ACTIONS_C16915,START_DATE,DATE,7,,Y
ACTIONS_C16915,COMPLETE_DATE,DATE,7,,Y
ACTIONS_C16915,ISTATUS,VARCHAR2,255,,Y
ACTIONS_C16915,ISTATUS_NAME,VARCHAR2,255,,Y
```
From what you said about the hex values, does that mean something is incorrect at the database level that would prevent it from being compared using the output with the DiffUtils library?

**Comment by wumpz** [comment]
No. The negative values were the first strange thing I saw. But somewhere between dB and string values must be the difference. You said that the hardcoded strings are processed correctly.

**Comment by akon89** [comment]
Yeah that's correct.  When I type them by hand, then they come out correctly, but when I get them from the database then there is a problem.  I was afraid I was doing something wrong with making the Lists that was making it not work with your library.  
I guess my actual code to put it into the List and the comparison are correct, but maybe there is something external to the program I wrote (like maybe some network thing?) that is throwing off the results.  I was thinking I had just done something wrong when I wrote the Strings out to the Lists and then ran it through the DiffUtils library.  As long as I have done that part correctly, I guess this would be something that is going wrong when the data is coming over from the database.
